genet/core.py

```diff
@@ -704,16 +693,40 @@ def add_nodes(self, nodes_and_attribs: dict, silent: bool = False, ignore_change
         reindexing_dict = {}
         if ('id' not in df_nodes.columns) or (df_nodes['id'].isnull().any()):
             df_nodes['id'] = df_nodes.index
-
+        if not {'lat', 'lon'}.issubset(set(df_nodes.columns)):
+            df_nodes[['lon', 'lat']] = float('nan')
+        if not {'x', 'y'}.issubset(set(df_nodes.columns)):
+            df_nodes[['x', 'y']] = float('nan')
+        if df_nodes[['lon', 'lat']].isnull().any().any():
+            missing_lat_lon = df_nodes[['lon', 'lat']].isnull().T.any()
+            df_nodes.loc[missing_lat_lon, ['lon', 'lat']] = pd.DataFrame(list(df_nodes.loc[missing_lat_lon].apply(
+                lambda row: spatial.change_proj(row['x'], row['y'], self.transformer), axis=1)),
+                columns=['lon', 'lat'], index=df_nodes.loc[missing_lat_lon].index
+            )
+        if df_nodes[['x', 'y']].isnull().any().any():
+            missing_x_y = df_nodes[['x', 'y']].isnull().T.any()
+            transformer = Transformer.from_crs('epsg:4326', self.epsg, always_xy=True)
+            df_nodes.loc[missing_x_y, ['x', 'y']] = pd.DataFrame(list(df_nodes.loc[missing_x_y].apply(
+                lambda row: spatial.change_proj(row['lon'], row['lat'], transformer), axis=1)),
+                columns=['x', 'y'], index=df_nodes.loc[missing_x_y].index
+            )
+
+        nodes_and_attribs_to_add = dict_support.merge_complex_dictionaries(
+            df_nodes[['x', 'y', 'lon', 'lat', 'id']].T.to_dict(), nodes_and_attribs)
+
+        # pandas is terrible with large numbers so we update them/generate them here
+        for node, attribs in nodes_and_attribs_to_add.items():
+            if 's2_id' not in nodes_and_attribs[node]:
+                attribs['s2_id'] = spatial.generate_index_s2(attribs['lat'], attribs['lon'])
```

Comment:
this fills in missing spatial info, x,y or lat,lon if only one is passed - which is preffered so people dont have to reproject anything themselves, and s2 ids are added for convenience / math with the rest of the network

Reply:
some really cool stuff @KasiaKoz ! 🥇